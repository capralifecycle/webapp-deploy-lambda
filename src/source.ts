import type * as iam from "aws-cdk-lib/aws-iam"
import type * as s3 from "aws-cdk-lib/aws-s3"
import * as s3Assets from "aws-cdk-lib/aws-s3-assets"
import type * as constructs from "constructs"

// This is mostly based on aws-s3-deployment from aws-cdk.

export interface SourceConfig {
  /**
   * The source bucket to deploy from.
   */
  readonly bucket: s3.IBucket

  /**
   * An S3 object key in the source bucket that points to a zip file.
   */
  readonly zipObjectKey: string
}

/**
 * Bind context for ISources
 */
export interface SourceContext {
  /**
   * The role for the handler
   *
   * @default - no policy is modified
   */
  readonly handlerRole?: iam.IRole
}

/**
 * Represents a source for bucket deployments.
 */
export interface ISource {
  /**
   * Binds the source to a bucket deployment.
   */
  bind(scope: constructs.Construct, context?: SourceContext): SourceConfig
}

/**
 * Specifies bucket deployment source.
 *
 * Usage:
 *
 *     Source.bucket(bucket, key)
 *     Source.asset('/local/path/to/directory')
 *     Source.asset('/local/path/to/a/file.zip')
 *
 */
export class Source {
  /**
   * Uses a .zip file stored in an S3 bucket as the source for the destination bucket contents.
   * @param bucket The S3 Bucket
   * @param zipObjectKey The S3 object key of the zip file with contents
   */
  public static bucket(bucket: s3.IBucket, zipObjectKey: string): ISource {
    return {
      bind: (_: constructs.Construct, context?: SourceContext) => {
        if (!context) {
          throw new Error("To use a Source.bucket(), context must be provided")
        }

        if (context.handlerRole) {
          bucket.grantRead(context.handlerRole)
        }

        return { bucket, zipObjectKey }
      },
    }
  }

  /**
   * Uses a local asset as the deployment source.
   * @param path The path to a local .zip file or a directory
   */
  public static asset(path: string, options?: s3Assets.AssetOptions): ISource {
    return {
      bind(scope: constructs.Construct, context?: SourceContext): SourceConfig {
        if (!context) {
          throw new Error("To use a Source.asset(), context must be provided")
        }

        let id = 1
        while (scope.node.tryFindChild(`Asset${id}`)) {
          id++
        }
        const asset = new s3Assets.Asset(scope, `Asset${id}`, {
          path,
          ...options,
        })
        if (!asset.isZipArchive) {
          throw new Error(
            "Asset path must be either a .zip file or a directory",
          )
        }

        if (context.handlerRole) {
          asset.grantRead(context.handlerRole)
        }

        return {
          bucket: asset.bucket,
          zipObjectKey: asset.s3ObjectKey,
        }
      },
    }
  }

  private constructor() {}
}
