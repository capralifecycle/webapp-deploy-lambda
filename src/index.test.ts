import * as cdk from "aws-cdk-lib"
import * as cloudfront from "aws-cdk-lib/aws-cloudfront"
import * as iam from "aws-cdk-lib/aws-iam"
import * as s3 from "aws-cdk-lib/aws-s3"
import "jest-cdk-snapshot"
import * as path from "path"
import { fileURLToPath } from "url"
import { WebappDeploy } from "."
import { Source } from "./source"

const __dirname = path.dirname(fileURLToPath(import.meta.url))

test("WebappDeploy", () => {
  const app = new cdk.App()
  const stack = new cdk.Stack(app, "Stack", {
    env: {
      account: "123456789123",
      region: "eu-west-1",
    },
  })

  const buildsBucket = s3.Bucket.fromBucketName(
    stack,
    "BuildsBucket",
    "builds-bucket",
  )
  const webBucket = s3.Bucket.fromBucketName(stack, "WebBucket", "web-bucket")

  const distribution = cloudfront.Distribution.fromDistributionAttributes(
    stack,
    "Distribution",
    {
      distributionId: "EKJ2IPY1KTEAR1",
      domainName: "example.com",
    },
  )

  const callerRole = new iam.Role(stack, "CallerRole", {
    assumedBy: new iam.AnyPrincipal(),
  })

  const webappDeploy = new WebappDeploy(stack, "WebappDeploy", {
    buildsBucket,
    distribution,
    webBucket,
  })

  webappDeploy.deployFn.grantInvoke(callerRole)

  expect(stack).toMatchCdkSnapshot()
})

test("WebappDeploy with source", () => {
  const app = new cdk.App()
  const stack = new cdk.Stack(app, "Stack", {
    env: {
      account: "123456789123",
      region: "eu-west-1",
    },
  })

  const webBucket = s3.Bucket.fromBucketName(
    stack,
    "WebBucket",
    "example-bucket",
  )

  const distribution = cloudfront.Distribution.fromDistributionAttributes(
    stack,
    "Distribution",
    {
      distributionId: "EKJ2IPY1KTEAR1",
      domainName: "example.com",
    },
  )

  new WebappDeploy(stack, "WebappDeploy", {
    distribution,
    source: Source.asset(path.join(__dirname, "../example-assets/source")),
    webBucket,
  })

  expect(stack).toMatchCdkSnapshot()
})
