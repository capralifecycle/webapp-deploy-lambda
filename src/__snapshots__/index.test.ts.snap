// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WebappDeploy 1`] = `
Object {
  "Resources": Object {
    "CallerRole7E4BBD59": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": "*",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "CallerRoleDefaultPolicyECDC9714": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "WebappDeploy4D661524",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CallerRoleDefaultPolicyECDC9714",
        "Roles": Array [
          Object {
            "Ref": "CallerRole7E4BBD59",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "WebappDeploy4D661524": Object {
      "DependsOn": Array [
        "WebappDeployServiceRoleDefaultPolicyE446530F",
        "WebappDeployServiceRole78F79841",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": "cdk-hnb659fds-assets-123456789123-eu-west-1",
          "S3Key": "92a9729ddc0d8f311e1b66f1d5811ff2d5f57ace5368c906a0929813cf79e86f.zip",
        },
        "Environment": Object {
          "Variables": Object {
            "CF_DISTRIBUTION_ID": "EKJ2IPY1KTEAR1",
            "DEPLOY_LOG_BUCKET_URL": "s3://web-bucket/deployments.log",
            "EXPIRE_SECONDS": "432000",
            "TARGET_BUCKET_URL": "s3://web-bucket/web",
          },
        },
        "Handler": "webapp_deploy.main.handler",
        "ReservedConcurrentExecutions": 1,
        "Role": Object {
          "Fn::GetAtt": Array [
            "WebappDeployServiceRole78F79841",
            "Arn",
          ],
        },
        "Runtime": "python3.8",
        "Timeout": 120,
      },
      "Type": "AWS::Lambda::Function",
    },
    "WebappDeployServiceRole78F79841": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "WebappDeployServiceRoleDefaultPolicyE446530F": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "cloudfront:CreateInvalidation",
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::web-bucket",
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::web-bucket/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::builds-bucket",
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::builds-bucket/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "WebappDeployServiceRoleDefaultPolicyE446530F",
        "Roles": Array [
          Object {
            "Ref": "WebappDeployServiceRole78F79841",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;

exports[`WebappDeploy with source 1`] = `
Object {
  "Resources": Object {
    "AWS679f53fac002430cb0da5b7982bd22872D164C4C": Object {
      "DependsOn": Array [
        "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": "cdk-hnb659fds-assets-123456789123-eu-west-1",
          "S3Key": "4a575666d1c2c6412590d2a56f328e040a81ad1ef59aecee31ae9b393d05f659.zip",
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 120,
      },
      "Type": "AWS::Lambda::Function",
    },
    "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "WebappDeploy4D661524": Object {
      "DependsOn": Array [
        "WebappDeployServiceRoleDefaultPolicyE446530F",
        "WebappDeployServiceRole78F79841",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": "cdk-hnb659fds-assets-123456789123-eu-west-1",
          "S3Key": "92a9729ddc0d8f311e1b66f1d5811ff2d5f57ace5368c906a0929813cf79e86f.zip",
        },
        "Environment": Object {
          "Variables": Object {
            "CF_DISTRIBUTION_ID": "EKJ2IPY1KTEAR1",
            "DEPLOY_LOG_BUCKET_URL": "s3://example-bucket/deployments.log",
            "EXPIRE_SECONDS": "432000",
            "TARGET_BUCKET_URL": "s3://example-bucket/web",
          },
        },
        "Handler": "webapp_deploy.main.handler",
        "ReservedConcurrentExecutions": 1,
        "Role": Object {
          "Fn::GetAtt": Array [
            "WebappDeployServiceRole78F79841",
            "Arn",
          ],
        },
        "Runtime": "python3.8",
        "Timeout": 120,
      },
      "Type": "AWS::Lambda::Function",
    },
    "WebappDeployCustomResourceCF959726": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "WebappDeployCustomResourceCustomResourcePolicyB85F7B61",
      ],
      "Properties": Object {
        "Create": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"service\\":\\"Lambda\\",\\"action\\":\\"invoke\\",\\"physicalResourceId\\":{\\"id\\":\\"webapp-deploy\\"},\\"parameters\\":{\\"FunctionName\\":\\"",
              Object {
                "Ref": "WebappDeploy4D661524",
              },
              "\\",\\"Payload\\":\\"{\\\\\\"artifactS3Url\\\\\\":\\\\\\"s3://cdk-hnb659fds-assets-123456789123-eu-west-1/48be1f8516ade54ba72e36a3f42565c479b1daa69b4b3872a6c5215ab64f1ef5.zip\\\\\\"}\\"}}",
            ],
          ],
        },
        "InstallLatestAwsSdk": false,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "AWS679f53fac002430cb0da5b7982bd22872D164C4C",
            "Arn",
          ],
        },
        "Update": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"service\\":\\"Lambda\\",\\"action\\":\\"invoke\\",\\"physicalResourceId\\":{\\"id\\":\\"webapp-deploy\\"},\\"parameters\\":{\\"FunctionName\\":\\"",
              Object {
                "Ref": "WebappDeploy4D661524",
              },
              "\\",\\"Payload\\":\\"{\\\\\\"artifactS3Url\\\\\\":\\\\\\"s3://cdk-hnb659fds-assets-123456789123-eu-west-1/48be1f8516ade54ba72e36a3f42565c479b1daa69b4b3872a6c5215ab64f1ef5.zip\\\\\\"}\\"}}",
            ],
          ],
        },
      },
      "Type": "Custom::AWS",
      "UpdateReplacePolicy": "Delete",
    },
    "WebappDeployCustomResourceCustomResourcePolicyB85F7B61": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "WebappDeploy4D661524",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "WebappDeployCustomResourceCustomResourcePolicyB85F7B61",
        "Roles": Array [
          Object {
            "Ref": "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "WebappDeployServiceRole78F79841": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "WebappDeployServiceRoleDefaultPolicyE446530F": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "cloudfront:CreateInvalidation",
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::example-bucket",
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::example-bucket/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::cdk-hnb659fds-assets-123456789123-eu-west-1",
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::cdk-hnb659fds-assets-123456789123-eu-west-1/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "WebappDeployServiceRoleDefaultPolicyE446530F",
        "Roles": Array [
          Object {
            "Ref": "WebappDeployServiceRole78F79841",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
